File: tests/termviz/test_pty.py
"""Real process/terminal acceptance for the dependency-free C workspace."""
from __future__ import annotations

import os
from pathlib import Path
import re
import signal
import sys

from pty_support import Session

ROOT = Path(__file__).resolve().parents[2]
BUILD = Path(os.environ.get("BUILD_DIR", str(ROOT / "build"))).resolve()
DEMO = str(BUILD / "termviz-workspace")
CHILD = str(BUILD / "test-workspace-child")
EVIDENCE = ROOT / "build" / "workspace-evidence"
EVIDENCE.mkdir(parents=True, exist_ok=True)


def workspace() -> None:
    s = Session([DEMO], 80, 24)
    try:
        s.until("ACTIVITY")
        s.pump(.1)
        assert s.pump(.2) == b"", "Unchanged demo frame emitted terminal data"
        before = s.screen.clears
        s.send("\tjjj")
        s.until("pane 3 / offset 3")
        s.send("\tjj")
        s.until("pane 4 / offset 2")
        assert "pane 3 / offset 3" in s.screen.text(), "Scroll state leaked between panes"
        s.send("\t")
        s.pump(.1)
        s.send("h")
        s.pump(.1)
        assert "renderer" not in s.screen.text(), "Tree branch did not collapse"
        s.send("l")
        s.until("renderer")
        s.send("\x1b[<0;19;3M\x1b[<32;30;3M\x1b[<0;30;3m")
        s.pump(.15)
        assert s.screen.cells[2][28][0] == "│", "Divider did not follow mouse drag"
        assert s.screen.clears == before, "Interaction cleared the whole screen"
        s.screen.save(EVIDENCE / "workspace-80x24.html")
        s.resize(40, 10)
        s.send("\t")
        s.pump(.15)
        assert "FOCUS" in s.screen.text() and "q quit" in s.screen.text()
        assert s.screen.overflow == 0
        s.screen.save(EVIDENCE / "workspace-40x10.html")
        s.resize(140, 40)
        s.until("INSPECT")
        assert s.screen.overflow == 0
        s.screen.save(EVIDENCE / "workspace-140x40.html")
        s.close(b"q")
    except BaseException:
        s.abort()
        raise
    print("PASS workspace: tree, independent scrolling, focus, mouse divider, incremental output, three sizes, terminal restoration")


def shell() -> None:
    s = Session([DEMO, "--shell"], 140, 40)
    try:
        s.until("TV$ ")
        s.send("printf '\\033[31mSHELL_%s\\033[0m\\n' OK; false; printf 'STATUS:%s\\n' \"$?\"\r")
        s.until("SHELL_OK")
        s.until("STATUS:1")
        pid_match = re.search(r"Child PID: (\d+)", s.screen.text())
        assert pid_match
        child_pid = int(pid_match.group(1))
        s.send("i=0; while [ $i -lt 700 ]; do printf 'HISTORY_%03d\\n' $i; i=$((i+1)); done\r")
        s.until("HISTORY_699")
        s.until("History: 512 / 512 rows")
        s.send("\x02[k")
        s.pump(.2)
        assert "offset 1" in s.screen.text()
        s.send("\x02]")
        s.send("i=0; while [ $i -lt 12 ]; do printf 'STREAM_%02d\\n' $i; i=$((i+1)); sleep 0.1; done\r")
        s.send("\x02\t")
        s.until("FOCUS pane 4")
        s.until("STREAM_11")
        s.send("\x02\t\x02\t")
        s.send(f"{CHILD}\r")
        s.until("ALT_READY")
        s.until("Screen: alternate")
        s.send("x")
        s.until("SAFE_OUTPUT")
        assert b"]52;" not in s.raw and b"SHOULD_NOT_ESCAPE" not in s.raw
        s.resize(80, 24)
        s.send("r")
        s.until("SIZE:")
        model_size = re.search(r"Terminal: (\d+)x(\d+)", s.screen.text())
        reported = re.search(r"SIZE:(\d+):(\d+)", s.screen.text())
        assert model_size and reported and model_size.groups() == tuple(reversed(reported.groups()))
        s.screen.save(EVIDENCE / "shell-80x24.html")
        s.resize(40, 10)
        s.send("r")
        s.until("SIZE:")
        assert s.screen.overflow == 0
        s.screen.save(EVIDENCE / "shell-40x10.html")
        s.resize(140, 40)
        s.send("r")
        s.until("ALT_READY")
        s.screen.save(EVIDENCE / "shell-140x40.html")
        s.send("q")
        s.until("AFTER_ALT")
        s.send("printf 'ALT_STATUS:%s\\n' \"$?\"\r")
        s.until("ALT_STATUS:7")
        s.send("exit 9\r")
        s.until("exited / status 9")
        try:
            os.kill(child_pid, 0)
        except ProcessLookupError:
            pass
        else:
            raise AssertionError("Exited shell was not reaped")
        s.close(b"\x02q")
    except BaseException:
        s.abort()
        raise
    print("PASS shell: commands/status, bounded history, focus during streaming, real fullscreen client, escape isolation, child resize/exit and restoration")


def interruption() -> None:
    s = Session([DEMO, "--shell"], 140, 40)
    try:
        s.until("TV$ ")
        s.send("sleep 30 & printf 'JOB_PID:'; jobs -p; wait\r")
        match = s.until_pattern(r"JOB_PID:(\d+)")
        job_pid = int(match.group(1))
        s.close(signum=signal.SIGTERM, expected=143)
        # Background jobs receive the terminal/session hangup. A zombie is no
        # longer executing; inspect its state where the platform exposes ps.
        import subprocess
        state = subprocess.run(["ps", "-o", "stat=", "-p", str(job_pid)], capture_output=True, text=True, check=False).stdout.strip()
        assert not state or state.startswith("Z"), f"Owned job remained live: {state}"
    except BaseException:
        s.abort()
        raise
    s = Session([DEMO, "--command", "/nonexistent/termviz-child"], 80, 24)
    try:
        s.until("exited / status 127")
        s.close(b"\x02q")
    except BaseException:
        s.abort()
        raise
    print("PASS lifecycle: interruption stops the owned shell/job; exec failure is visible; terminal restored")


def hydra() -> None:
    argv = [str(BUILD / "hydra-tui"), "--hydra", str(ROOT / "tests/fixtures/tui/fake-hydra.sh"), "--theme", "dark"]
    s = Session(argv, 140, 40)
    try:
        s.until("HYDRA WORKSPACE")
        s.until("Observed: LIVE")
        before = s.screen.clears
        s.send("j")
        s.until("Observed: STALE")
        s.send("\tjjj\tjj")
        s.pump(.2)
        assert "scroll 3" in s.screen.text() and "scroll 2" in s.screen.text()
        s.send("\x1b[<0;42;4M\x1b[<32;50;4M\x1b[<0;50;4m")
        s.pump(.3)
        assert s.screen.clears == before
        s.screen.save(EVIDENCE / "hydra-140x40.html")
        s.resize(80, 24)
        s.until("NAVIGATION")
        assert s.screen.overflow == 0
        s.screen.save(EVIDENCE / "hydra-80x24.html")
        s.resize(40, 10)
        s.until("FOCUS")
        assert "q quit" in s.screen.text() and s.screen.overflow == 0
        s.screen.save(EVIDENCE / "hydra-40x10.html")
        s.close(b"q")
    except BaseException:
        s.abort()
        raise
    print("PASS Hydra: real adapter rows, selection/details, independent pane scroll, mouse divider, resize and restoration")


if __name__ == "__main__":
    try:
        workspace()
        shell()
        interruption()
        hydra()
    except AssertionError as error:
        print(f"FAIL {error}", file=sys.stderr)
        raise

File: tests/termviz/test_pty.c
#define _XOPEN_SOURCE 700
#include "pty_support.h"
#include <errno.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
static char root[4096], build[4096], demo[4096], child[4096], evidence[4096];
static void observer_resize(void) {
    const unsigned char frame[] = "\033[2J\033[H界";
    size_t split;
    for (split = 1; split < sizeof(frame) - 1; split++) {
        struct tv_screen s = {0};
        tv_screen_reset(&s, 80, 24, false);
        tv_screen_feed(&s, frame, split);
        tv_screen_reset(&s, 40, 10, true);
        tv_screen_feed(&s, frame + split, sizeof(frame) - 1 - split);
        if (split < 4)
            CHECK(!s.awaiting_clear && !strcmp(s.cells[0].text, "界"),
                  "resize completes a split clear sequence");
        tv_screen_feed(&s, frame, sizeof(frame) - 1);
        CHECK(!s.awaiting_clear && !s.overflow && !strcmp(s.cells[0].text, "界"),
              "resize preserves split UTF-8 and CSI streams");
        free(s.cells);
        free(s.text);
    }
    puts("PASS observer: resize at every UTF-8 and CSI byte boundary");
}
static void save(struct tv_session *s, const char *name) {
    char path[4096];
    tv_format(path, sizeof(path), "%s/%s", evidence, name);
    tv_save(s, path);
}
static void workspace(void) {
    struct tv_session s;
    int before;
    const char *argv[] = {demo, NULL};
    tv_open(&s, argv, 80, 24, NULL);
    tv_until(&s, "ACTIVITY", 3);
    tv_pump(&s, .1);
    CHECK(tv_pump(&s, .2) == 0, "unchanged frame emitted bytes");
    before = s.screen.clears;
    tv_send(&s, "\tjjj");
    tv_until(&s, "pane 3 / offset 3", 3);
    tv_send(&s, "\tjj");
    tv_until(&s, "pane 4 / offset 2", 3);
    CHECK(tv_contains(&s, "pane 3 / offset 3"), "independent pane scroll");
    tv_send(&s, "\t");
    tv_pump(&s, .1);
    tv_send(&s, "h");
    tv_pump(&s, .1);
    CHECK(!tv_contains(&s, "renderer"), "collapsed branch");
    tv_send(&s, "l");
    tv_until(&s, "renderer", 3);
    tv_send(&s, "\033[<0;19;3M\033[<32;30;3M\033[<0;30;3m");
    tv_pump(&s, .15);
    CHECK(!strcmp(s.screen.cells[2 * s.screen.cols + 28].text, "│"), "mouse divider position");
    CHECK(s.screen.clears == before, "incremental interaction");
    save(&s, "workspace-80x24.html");
    tv_resize(&s, 40, 10);
    tv_send(&s, "\t");
    tv_pump(&s, .15);
    CHECK(tv_contains(&s, "FOCUS") && tv_contains(&s, "q quit"), "small workspace controls");
    CHECK(!s.screen.overflow, "40x10 overflow");
    save(&s, "workspace-40x10.html");
    tv_resize(&s, 140, 40);
    tv_until(&s, "INSPECT", 3);
    CHECK(!s.screen.overflow, "140x40 overflow");
    save(&s, "workspace-140x40.html");
    tv_close(&s, "q", 0, 0);
    puts("PASS workspace: tree, independent scroll, focus, mouse, incremental output, three sizes, "
         "termios");
}
static void shell_test(void) {
    struct tv_session s;
    char groups[2][256], model[2][256], reported[2][256], command[8192];
    pid_t pid;
    const char *argv[] = {demo, "--shell", NULL};
    tv_open(&s, argv, 140, 40, NULL);
    tv_until(&s, "TV$ ", 3);
    tv_send(&s, "printf '\\033[31mSHELL_%s\\033[0m\\n' OK; false; printf 'STATUS:%s\\n' \"$?\"\r");
    tv_until(&s, "SHELL_OK", 3);
    tv_until(&s, "STATUS:1", 3);
    CHECK(tv_match(&s, "Child PID: ([0-9]+)", groups, 1), "shell PID");
    pid = (pid_t)atoi(groups[0]);
    tv_send(&s, "i=0; while [ $i -lt 700 ]; do printf 'HISTORY_%03d\\n' $i; i=$((i+1)); done\r");
    tv_until(&s, "HISTORY_699", 3);
    tv_until(&s, "History: 512 / 512 rows", 3);
    tv_send(&s, "\002[k");
    tv_pump(&s, .2);
    CHECK(tv_contains(&s, "offset 1"), "history scroll");
    tv_send(&s, "\002]");
    tv_send(
        &s,
        "i=0; while [ $i -lt 12 ]; do printf 'STREAM_%02d\\n' $i; i=$((i+1)); sleep 0.1; done\r");
    tv_send(&s, "\002\t");
    tv_until(&s, "FOCUS pane 4", 3);
    tv_until(&s, "STREAM_11", 3);
    tv_send(&s, "\002\t\002\t");
    tv_format(command, sizeof(command), "%s\r", child);
    tv_send(&s, command);
    tv_until(&s, "ALT_READY", 3);
    tv_until(&s, "Screen: alternate", 3);
    tv_send(&s, "x");
    tv_until(&s, "SAFE_OUTPUT", 3);
    CHECK(!strstr(s.raw, "]52;") && !strstr(s.raw, "SHOULD_NOT_ESCAPE"), "escape isolation");
    tv_resize(&s, 80, 24);
    tv_send(&s, "r");
    tv_until(&s, "SIZE:", 3);
    CHECK(tv_match(&s, "Terminal: ([0-9]+)x([0-9]+)", model, 2) &&
              tv_match(&s, "SIZE:([0-9]+):([0-9]+)", reported, 2),
          "size reports");
    CHECK(!strcmp(model[0], reported[1]) && !strcmp(model[1], reported[0]), "child/model geometry");
    save(&s, "shell-80x24.html");
    tv_resize(&s, 40, 10);
    tv_send(&s, "r");
    tv_until(&s, "SIZE:", 3);
    CHECK(!s.screen.overflow, "small shell overflow");
    save(&s, "shell-40x10.html");
    tv_resize(&s, 140, 40);
    tv_send(&s, "r");
    tv_until(&s, "ALT_READY", 3);
    save(&s, "shell-140x40.html");
    tv_send(&s, "q");
    tv_until(&s, "AFTER_ALT", 3);
    tv_send(&s, "printf 'ALT_STATUS:%s\\n' \"$?\"\r");
    tv_until(&s, "ALT_STATUS:7", 3);
    tv_send(&s, "exit 9\r");
    tv_until(&s, "exited / status 9", 3);
    CHECK(kill(pid, 0) < 0 && errno == ESRCH, "shell reaped");
    tv_close(&s, "\002q", 0, 0);
    puts("PASS shell: command status, bounded history, stream focus, alternate client, escape "
         "isolation, resize, reap");
}
static void interruption(void) {
    struct tv_session s;
    char groups[1][256], out[1024];
    const char *argv[] = {demo, "--shell", NULL};
    tv_open(&s, argv, 140, 40, NULL);
    tv_until(&s, "TV$ ", 3);
    tv_send(&s, "sleep 30 & printf 'JOB_PID:'; jobs -p; wait\r");
    tv_until_match(&s, "JOB_PID:([0-9]+)", groups, 1, 3);
    tv_close(&s, NULL, SIGTERM, 143);
    {
        const char *ps[] = {"ps", "-o", "stat=", "-p", groups[0], NULL};
        char *p = out;
        (void)tv_command(NULL, NULL, out, sizeof(out), 5, ps);
        while (*p == ' ' || *p == '\n')
            p++;
        CHECK(!*p || *p == 'Z', "owned job stopped");
    }
    {
        const char *bad[] = {demo, "--command", "/nonexistent/termviz-child", NULL};
        tv_open(&s, bad, 80, 24, NULL);
        tv_until(&s, "exited / status 127", 3);
        tv_close(&s, "\002q", 0, 0);
    }
    puts("PASS lifecycle: interruption, owned job termination, visible exec failure, termios");
}
static void hydra(void) {
    struct tv_session s;
    char tui[4096], fake[4096];
    int before;
    tv_format(tui, sizeof(tui), "%s/hydra-tui", build);
    tv_format(fake, sizeof(fake), "%s/tests/fixtures/tui/fake-hydra.sh", root);
    {
        const char *argv[] = {tui, "--hydra", fake, "--theme", "dark", NULL};
        tv_open(&s, argv, 140, 40, NULL);
    }
    tv_until(&s, "HYDRA WORKSPACE", 3);
    tv_until(&s, "Observed: LIVE", 3);
    before = s.screen.clears;
    tv_send(&s, "j");
    tv_until(&s, "Observed: STALE", 3);
    tv_send(&s, "\tjjj\tjj");
    tv_pump(&s, .2);
    CHECK(tv_contains(&s, "scroll 3") && tv_contains(&s, "scroll 2"), "Hydra independent scroll");
    tv_send(&s, "\033[<0;42;4M\033[<32;50;4M\033[<0;50;4m");
    tv_pump(&s, .3);
    CHECK(s.screen.clears == before, "Hydra incremental drag");
    save(&s, "hydra-140x40.html");
    tv_resize(&s, 80, 24);
    tv_until(&s, "NAVIGATION", 3);
    CHECK(!s.screen.overflow, "Hydra medium overflow");
    save(&s, "hydra-80x24.html");
    tv_resize(&s, 40, 10);
    tv_until(&s, "FOCUS", 3);
    CHECK(tv_contains(&s, "q quit") && !s.screen.overflow, "Hydra small frame");
    save(&s, "hydra-40x10.html");
    tv_close(&s, "q", 0, 0);
    puts("PASS Hydra: adapter rows, selection, pane scroll, divider, resize and restoration");
}
static void hydra_startup_open(struct tv_session *s, const char *tui, const char *fake) {
    const char *argv[] = {tui, "--hydra", fake, NULL};
    tv_open(s, argv, 140, 40, NULL);
    tv_until(s, "Loading snapshot...", 3);
    tv_until(s, "visualization-proof", 3);
    CHECK(tv_contains(s, "No matching head"), "workflow arrives before delayed heads");
}
static void hydra_startup_repopulation(struct tv_session *s, const char *model, const char *data) {
    tv_send(s, "C");
    tv_until(s, "Observed: LIVE", 3);
    CHECK(tv_contains(s, "*  └ feature-live"), "first head selected after mode input");
    tv_send(s, "Aj");
    tv_until(s, "Observed: STALE", 3);
    CHECK(tv_contains(s, ">  └ feature-stale"), "single move selects second head");
    tv_write(model, "HYDRA_TUI\t2\n");
    tv_until(s, "No matching head", 3);
    tv_write(model, data);
    tv_until(s, "Observed: LIVE", 3);
    CHECK(tv_contains(s, ">- Current project"), "repopulation does not reseed selection");
}
static void hydra_startup_selection(void) {
    struct tv_session s;
    char tui[4096], fake[4096], source[4096], model[4096], data[8192], name[64];
    const char *keys[] = {"k", "j", "\r"};
    const char *selected[] = {">- Current project", ">  └ visualization-proof", ">+ Current project"};
    size_t i;
    tv_format(tui, sizeof(tui), "%s/hydra-tui", build);
    tv_format(fake, sizeof(fake), "%s/tests/fixtures/tui/fake-hydra.sh", root);
    tv_format(source, sizeof(source), "%s/tests/fixtures/tui/native-v2.tsv", root);
    tv_format(model, sizeof(model), "%s/startup-model.tsv", evidence);
    tv_read(source, data, sizeof(data));
    CHECK(!setenv("HYDRA_TEST_TUI_DELAY", "1", 1) && !setenv("HYDRA_TUI_FIXTURE", model, 1),
          "configure delayed startup snapshot");
    tv_write(model, data);
    hydra_startup_open(&s, tui, fake);
    hydra_startup_repopulation(&s, model, data);
    save(&s, "startup-repopulation.html");
    tv_close(&s, "q", 0, 0);
    for (i = 0; i < sizeof(keys) / sizeof(keys[0]); i++) {
        hydra_startup_open(&s, tui, fake);
        tv_send(&s, keys[i]);
        tv_until(&s, "Observed: LIVE", 3);
        CHECK(tv_contains(&s, selected[i]), "explicit root/run/collapse retained after heads arrive");
        tv_format(name, sizeof(name), "startup-selection-%zu.html", i);
        save(&s, name);
        tv_close(&s, "q", 0, 0);
    }
    CHECK(!unsetenv("HYDRA_TEST_TUI_DELAY") && !unsetenv("HYDRA_TUI_FIXTURE"),
          "restore startup fixture environment");
    puts("PASS Hydra startup: first head, single move, explicit root/run/collapse, repopulation");
}
int main(int argc, char **argv) {
    tv_init();
    observer_resize();
    tv_paths(root, sizeof(root), build, sizeof(build));
    tv_format(demo, sizeof(demo), "%s/termviz-workspace", build);
    tv_format(child, sizeof(child), "%s/test-workspace-child", build);
    tv_format(evidence, sizeof(evidence), "%s/workspace-evidence", build);
    tv_mkdir(evidence);
    CHECK(argc == 1 || (argc == 2 && !strcmp(argv[1], "--standalone")),
          "usage: test-pty [--standalone]");
    workspace();
    shell_test();
    interruption();
    if (argc == 1) {
        hydra();
        hydra_startup_selection();
    }
    return 0;
}
